### DISCUSSION TITLE
Recommended way to compile contracts with custom settings on the fly

### DISCUSSION THREAD

**Discussion by mds1** [NomicFoundation/hardhat#2093] [General]
I'm trying to write a method that takes a contract address as input, fetches the source code from [Etherscan](https://docs.etherscan.io/api-endpoints/contracts#get-contract-source-code-for-verified-contract-source-codes), and uses the returned source code + compiler settings to re-compile the source code with the `storageLayout` flag (and possibly other solc flags) enabled.

A sample response from this Etherscan endpoint can be found by querying `https://api.etherscan.io/api?module=contract&action=getsourcecode&address=0xc0Da02939E1441F497fd74F78cE7Decb17B66529&apikey=yourEtherscanApiKey` (make sure to fill in your API key)

The source code can be flattened into a single string, and import statements and duplicate pragmas can then be removed. From there, you can save the source code to a temporary file in the `contracts` directory, then the final step is to run the compile task with the settings returned from Etherscan plus any additional settings.

That final step is what I'd like to discuss here. @fvictorio mentioned in Discord that there's currently no way to do something like `await run('compile', myCustomSolcSettingsObject)`, so I wanted to get thoughts on the best way to do this. It seems I should be able to import the `Compiler` class from [here](https://github.com/nomiclabs/hardhat/tree/master/packages/hardhat-core/src/internal/solidity/compiler) and use that to compile with custom settings, so that's what I'm planning to try next, but would love to hear any other thoughts here!

**Comment by alcuadrado** [comment]
Hey @mds1,

Something that you can do is using the subtasks of `compile`, creating a `CompilationJob` and compiling that. You'd be reimplementing some parts of the compilation pipeline, but those would be rather small.

**Reply by mds1** [reply]
Thanks @alcuadrado! Do you mind expanding on what you mean a bit? 

Are you saying to write my own compilation task that effectively re-implements the `TASK_COMPILE_SOLIDITY` [task](https://github.com/nomiclabs/hardhat/blob/6aa32687af6bf1edd10fe23fc32a96d9c50da1f1/packages/hardhat-core/src/builtin-tasks/compile.ts#L1265-L1371), and just allow the `solcConfig` on L1348 to be specified by an input parameter to this task?

That feels like a good chunk of code to copy/paste, so I feel like I may be misunderstanding something

**Reply by alcuadrado** [reply]
Now that I think about this again, you should be able to override `TASK_COMPILE_SOLIDITY_GET_COMPILATION_JOBS` and add your own custom `CompilationJob`s on top of the default ones, without having to copy/paste code
